Machine Learning Assignment 3
by Ming Nie mn2984

In [1]:
#Import Python Libraries
import numpy as np
import scipy as sp
import pandas as pd
import matplotlib as mpl
import seaborn as sns
import statistics

Homework Instructions:

Your final document should be an html file generated from a Jupyter notebook.   (i.e.-save your notebook via file>download as>html).  The file should be uploaded to this assignment on the course website.  

In answering each of the following questions please include a) the question as a markdown header in your Jupyter notebook, b)  the raw code that you used to generate any results, tables, or figures, and c) the top ten or fewer rows of the dataframe (do not include more than ten rows for any table in your report).

Include any plots or figures generated from your code as well.

Homework Questions:

Unsupervised learning for red wine characteristics

First, import the red wine csv file into a pandas dataframe from the following website:

https://archive.ics.uci.edu/ml/machine-learning-databases/wine-quality/

(Note: you need to adjust the argument for read_csv() from sep=',' to sep=';')

Use the wine quality data set to answer the following questions. Present any figures that assist you in your analysis:



In [2]:
df = pd.read_csv(r"C:\Users\14249\OneDrive\Columbia\$ML\Assignment3\winequality-red.csv", sep = ";")
df.head(10)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   
5            7.4              0.66         0.00             1.8      0.075   
6            7.9              0.60         0.06             1.6      0.069   
7            7.3              0.65         0.00             1.2      0.065   
8            7.8              0.58         0.02             2.0      0.073   
9            7.5              0.50         0.36             6.1      0.071   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   
5                 13.0                  40.0   0.9978  3.51       0.56   
6                 15.0                  59.0   0.9964  3.30       0.46   
7                 15.0                  21.0   0.9946  3.39       0.47   
8                  9.0                  18.0   0.9968  3.36       0.57   
9                 17.0                 102.0   0.9978  3.35       0.80   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5  
5      9.4        5  
6      9.4        5  
7     10.0        7  
8      9.5        7  
9     10.5        5

1. Use K Means Cluster Analysis to identify cluster(s) of observations that have high and low values of the wine quality. (Assume all variables are continuous.)

Describe variables that cluster with higher values of wine quality. Describe variables that cluster with lower values of wine quality.

Answer:
It seems that high quality requires high level in alcohol, low level in volatile acidity. Low quality wine usually has low level in alcohol, high level in volatile acidity.

If you want to make a good bottle of wine, then what characteristics are most important according to this analysis? 

Answer: I would say alcohol is the most important one. Since the two groups in good quality are quite high in alcohol level. It is the variable that good wine and bad wine differentiate the most.

In [3]:
from sklearn.cluster import KMeans
from sklearn.preprocessing import scale
from sklearn.decomposition import PCA

In [4]:
X = pd.DataFrame(scale(df), index=df.index, columns=df.columns)
km1 = KMeans(n_clusters=5, n_init=20)
km1.fit(X)


KMeans(n_clusters=5, n_init=20)

In [5]:
X['cluster'] = km1.labels_
display(X.groupby('cluster').mean().sort_values('quality'))

fixed acidity  volatile acidity  citric acid  residual sugar  \
cluster                                                                 
2            -0.483685          0.675532    -0.791517       -0.231587   
0            -0.078895          0.085668     0.098161        0.433196   
3             0.081831          0.017955     1.144178       -0.399396   
4             1.367270         -0.674977     1.140048        0.198292   
1            -0.599699         -0.485038    -0.108176       -0.221834   

         chlorides  free sulfur dioxide  total sulfur dioxide   density  \
cluster                                                                   
2        -0.078366            -0.401776             -0.384522 -0.095559   
0         0.037128             1.085767              1.381157  0.384663   
3         5.604731            -0.070479              0.474416  0.185803   
4        -0.014903            -0.538583             -0.523448  0.833230   
1        -0.380796             0.184010             -0.201640 -1.139002   

               pH  sulphates   alcohol   quality  
cluster                                           
2        0.409664  -0.419668 -0.425757 -0.465275  
0       -0.156750  -0.184107 -0.618165 -0.463049  
3       -1.687357   3.720608 -0.882563 -0.360695  
4       -0.841194   0.330386  0.172104  0.402409  
1        0.528551   0.200265  1.215054  0.832554

2. Use Hierarchical Cluster Analysis to identify cluster(s) of observations that have high and low values of the wine quality. (Assume all variables are continuous.) Use complete linkage and the same number of groups that you found to be the most meaningful in question 1.

Describe variables that cluster with higher values of wine quality. Describe variables that cluster with lower values of wine quality.

Answer:
The variables that cluster with higher values of wine quality: high level in alcohol. 

The variables that cluster with lower values of wine quality: low level in alcohol.

If you want to make a good bottle of wine, then what characteristics are most important according to this analysis? Have your conclusions changed using Hierarchical clustering rather than k means clustering? Present any figures that assist you in your analysis.

Answer: the most important one is the high level in alcohol level. It is the variable that good wine and bad wine differentiate the most. The conclusion remains the same. The dataframe is sorted by quality in ascending order, the conclusion is drawn based on the last two columns.

In [6]:
from sklearn.cluster import AgglomerativeClustering

X_2 = pd.DataFrame(scale(df), index=df.index, columns=df.columns)

ac = AgglomerativeClustering(n_clusters=5,   
                             affinity='euclidean', 
                             linkage='complete') 
X_2['cluster_hierarchical'] = ac.fit_predict(X)
display(X_2.groupby('cluster_hierarchical').mean().sort_values('quality'))

fixed acidity  volatile acidity  citric acid  \
cluster_hierarchical                                                 
3                          0.074897         -0.350945     3.127368   
1                         -0.407253          0.219714    -0.380440   
2                          0.338224         -0.648891     0.641150   
0                          1.244709         -0.609158     1.103051   
4                         -0.539850         -1.512932     0.816598   

                      residual sugar  chlorides  free sulfur dioxide  \
cluster_hierarchical                                                   
3                           0.043416  11.116408             0.394484   
1                          -0.097163  -0.136166             0.133354   
2                           7.303740   0.668432             1.350792   
0                           0.062475   0.358610            -0.505035   
4                           0.320113  -0.826059             2.130183   

                      total sulfur dioxide   density        pH  sulphates  \
cluster_hierarchical                                                        
3                                 0.320274  0.770280 -2.663717   5.735193   
1                                 0.110770 -0.244611  0.315549  -0.198955   
2                                 0.842294  1.950444 -0.606546   0.084691   
0                                -0.446908  0.747608 -0.894324   0.600903   
4                                 3.172635 -2.565626 -1.374340  -1.027706   

                       alcohol   quality  
cluster_hierarchical                      
3                    -0.960246 -1.407158  
1                    -0.031163 -0.114178  
2                    -0.608244  0.037958  
0                     0.085244  0.326663  
4                     1.301955  1.070184

3. Use Principal Components Analysis to reduce the dimensions of your data. 

How much of the variation in your data is explained by the first two principal components. 

Answer: 
The amount of variance explained by each of the selected components: 5.3664
Percentage of variance explained by each of the selected components: 0.4469


How might you use the first two components to do supervised learning on some other variable tied to wine (e.g. - wine price)?

Answer: We can use the data of the first two compoenents to run a lasso regression to predict other variable tied to wine. In the case of the wine price, we can set the wine price as Y, and the two components as X in our lasso regression model.

In [7]:
from sklearn.decomposition import PCA

X_3 = pd.DataFrame(scale(df), index=df.index, columns=df.columns)
pca = PCA()
pca.fit_transform(X_3)
pca.explained_variance_


array([3.12312087, 2.24328497, 1.68397283, 1.21578121, 0.97387267,
       0.66300688, 0.61870473, 0.50618912, 0.41156493, 0.3281246 ,
       0.18033141, 0.05955516])

In [8]:
pca.explained_variance_ratio_


array([0.26009731, 0.1868235 , 0.14024331, 0.10125174, 0.0811053 ,
       0.05521602, 0.05152648, 0.04215605, 0.03427563, 0.02732662,
       0.01501822, 0.00495983])

In [9]:
pca.explained_variance_[0] + pca.explained_variance_[1]

5.366405841697327

In [10]:
pca.explained_variance_ratio_[0] + pca.explained_variance_ratio_[1]

0.4469208117069171